vertical-pod-autoscaler/pkg/admission-controller/logic/limitrange_checker.go

```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logic
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+
+	v1_listers "k8s.io/client-go/listers/core/v1"
+)
+
+// LimitsHints provide hinted limits that respect limit range max ratio
+type LimitsHints interface {
+	IsNil() bool
+	RequestsExceedsRatio(indexOfContainer int, resourceName v1.ResourceName) bool
+	HintedLimit(indexOfContainer int, resourceName v1.ResourceName) resource.Quantity
+}
+
+// LimitRangeHints implements LimitsHints interface
+type LimitRangeHints struct {
+	requestsExceedsRatio  []map[v1.ResourceName]bool
+	limitsRespectingRatio []v1.ResourceList
+}
+
+var _ LimitsHints = &LimitRangeHints{}
+
+// LimitsChecker checks for LimitRange and if container needs limits to be set
+type LimitsChecker interface {
+	NeedsLimits(*v1.Pod, []ContainerResources) LimitsHints
+}
+
+// IsNil return true if there is no hints to set limits
+func (lrh *LimitRangeHints) IsNil() bool {
+	return lrh == (*LimitRangeHints)(nil)
+}
+
+// RequestsExceedsRatio return true if limits have to be set to respect limit range max ratio
+func (lrh *LimitRangeHints) RequestsExceedsRatio(indexOfContainer int, resourceName v1.ResourceName) bool {
+	if !lrh.IsNil() {
+		yes, ok := lrh.requestsExceedsRatio[indexOfContainer][resourceName]
+		return ok && yes
+	}
+	return false
+}
+
+// HintedLimit return the limit Quantity that respect the limit range max ration
+func (lrh *LimitRangeHints) HintedLimit(indexOfContainer int, resourceName v1.ResourceName) resource.Quantity {
+	if !lrh.IsNil() {
+		limit, ok := lrh.limitsRespectingRatio[indexOfContainer][resourceName]
+		if ok {
+			return limit
+		}
+		return resource.Quantity{}
+	}
+	return resource.Quantity{}
+}
+
+type neverNeedsLimitsChecker struct{}
+
+var _ LimitsChecker = &neverNeedsLimitsChecker{}
+
+func (lc *neverNeedsLimitsChecker) NeedsLimits(pod *v1.Pod, containersResources []ContainerResources) LimitsHints {
+	return LimitsHints((*LimitRangeHints)(nil))
```

Comment:
It might be simpler if we returned an object with two empty lists instead of nil.